Documentation for index-manipulation and contraction API rework (#438)
Jutho/TensorKit.jl: docs/src/man/contractions.md

# [Tensor contractions and tensor networks](@id ss_tensor_contraction)

> One of the most important operation with tensor maps is to compose them, more generally known as contracting them.

Reviewer: ```suggestion
One of the most important operations with tensor maps is to compose them, more generally known as contracting them.
```

Technically, we can embed this diagram in ``ℝ × [0,1]`` and attach all the unconnected line endings corresponding objects in the source at some position ``(x,0)`` for ``x∈ℝ``, and all line endings corresponding to objects in the target at some position ``(x,1)``.
The resulting morphism is then invariant under what is known as *framed three-dimensional isotopy*, i.e. three-dimensional rearrangements of the morphism that respect the rules of boxes connected by ribbons whose open endings are kept fixed.
Such a two-dimensional diagram cannot easily be encoded in a single line of code.

However, things simplify when the braiding is symmetric (such that over- and under- crossings become equivalent, i.e. just crossings), and when twists, i.e. self-crossings in this case, are trivial.
This amounts to `BraidingStyle(I) == Bosonic()` in the language of TensorKit.jl, and is true for any subcategory of ``\mathbf{Vect}``, i.e. ordinary tensors, possibly with some symmetry constraint.

> The case of ``\mathbf{SVect}`` and its subcategories, and more general categories, are discussed below.

Reviewer: Personally, I feel like the category theory language used here is overkill, in the sense that it's overly technical. It also refers to a section which we explicitly say is optional to go through.

In the case of trivial twists, we can deform the diagram such that we first combine every morphism with a number of coevaluations ``η`` so as to represent it as a tensor, i.e. with a trivial domain.

> We can then rearrange the morphism to be all ligned up horizontally, where the original morphism compositions are now being performed by evaluations ``ϵ``.

Reviewer: ```suggestion
We can then rearrange the morphism to be all lined up horizontally, where the original morphism compositions are now being performed by evaluations ``ϵ``.
```

This process will generate a number of crossings and twists, where the latter can be omitted because they act trivially.
Similarly, double crossings can also be omitted.

> As a consequence, the diagram, or the morphism it represents, is completely specified by the tensors it is composed of, and which indices between the different tensors are connect, via the evaluation ``ϵ``, and which indices make up the source and target of the resulting morphism.

Reviewer: ```suggestion
As a consequence, the diagram, or the morphism it represents, is completely specified by the tensors it is composed of, and which indices between the different tensors are connected, via the evaluation ``ϵ``, and which indices make up the source and target of the resulting morphism.
```

If we also compose the resulting morphisms with coevaluations so that it has a trivial domain, we just have one type of unconnected lines, henceforth called open indices.

> We sketch such a rearrangement in the following picture

Reviewer: ```suggestion
We sketch such a rearrangement in the following picture:
```

For this, we rely on the interface provided by the package [TensorOperations.jl](https://github.com/QuantumKitHub/TensorOperations.jl).

The above picture would be encoded as
```julia
@tensor E[a, b, c, d, e] := A[v, w, d, x] * B[y, z, c, x] * C[v, e, y, b] * D[a, w, z]
```
or
```julia
@tensor E[:] := A[1, 2, -4, 3] * B[4, 5, -3, 3] * C[1, -5, 4, -2] * D[-1, 2, 5]
```
where the latter syntax is known as NCON-style, and labels the unconnected or outgoing indices with negative integers, and the contracted indices with positive integers.

A number of remarks are in order.
TensorOperations.jl accepts both integers and any valid variable name as dummy label for indices, and everything in between `[ ]` is not resolved in the current context but interpreted as a dummy label.
Here, we label the indices of a `TensorMap`, like `A::TensorMap{T, S, N₁, N₂}`, in a linear fashion, where the first position corresponds to the first space in `codomain(A)`, and so forth, up to position `N₁`.
Index `N₁ + 1` then corresponds to the first space in `domain(A)`.

> However, because we have applied the coevaluation ``η``, it actually corresponds to the corresponding dual space, in accordance with the interface of [`space(A, i)`](@ref) that we introduced [above](@ref ss_tensor_properties), and as indiated by the dotted box around ``A`` in the above picture.

Reviewer 2: ```suggestion
However, because we have applied the coevaluation ``η``, it actually corresponds to the corresponding dual space, in accordance with the interface of [`space(A, i)`](@ref), and as indiated by the dotted box around ``A`` in the above picture.
```

Note that for existing tensors, both on the left hand side and right hand side, trying to specify the indices in the domain and the codomain seperately using the above syntax, has no effect, as the bipartition of indices are already fixed by the existing object.
Hence, if `E` has been created by the previous line of code, all of the following lines are now equivalent
```julia
@tensor E[(a, b, c);(d, e)] = A[v, w, d, x] * B[y, z, c, x] * C[v, e, y, b] * D[a, w, z]
@tensor E[a, b, c, d, e] = A[v w d; x] * B[(y, z, c); (x, )] * C[v e y; b] * D[a, w, z]
@tensor E[a b; c d e] = A[v; w d x] * B[y, z, c, x] * C[v, e, y, b] * D[a w; z]
```
and none of those will or can change the partition of the indices of `E` into its codomain and its domain.

Two final remarks are in order.
Firstly, the order of the tensors appearing on the right hand side is irrelevant, as we can reorder them by using the allowed moves of the Penrose graphical calculus, which yields some crossings and a twist.
As the latter is trivial, it can be omitted, and we just use the same rules to evaluate the newly ordered tensor network.

> For the particular case of matrix-matrix multiplication, which also captures more general settings by appropriotely combining spaces into a single line, we indeed find

Reviewer 2: ```suggestion
For the particular case of matrix-matrix multiplication, which also captures more general settings by appropriately combining spaces into a single line, we indeed find
```

```@example fermioncontraction
AX = repartition(A, 2, 1) * X
AXB = repartition(AX, 1, 2) * B
```

Alternatively, we could decide that we first wish to contract `A` with `B`, and only then contract the result with `X`:

```@example fermioncontraction
AB = permute(A, ((1, 2), (3,))) * permute(B, ((2,), (1, 3)))
ABX = repartition(permute(AB, ((1, 4), (2, 3))) * repartition(X, 2, 0), 1, 1)
```

This is where the issue becomes clearer, as the results are no longer equal:

```@example fermioncontraction
AXB ≈ ABX
```

### Trivializing the twist

So what happened?
If we carefully inspect what we actually computed here, we can show that in order to deform one diagram into the other, we have to introduce a self-crossing, which then altered the result.
While the example here is still simple to follow, in general we would like that the result of `@tensor` expressions does not depend on the input order of the tensors.
This is especially true for larger expressions where we wish to dynamically compute the optimal contraction order, as this would alter the order in a very non-transparent manner.

The way out of this effectively consists of absorbing this twist in the coevaluation map ``η``.

> This modified map ``̃η := η ∘ θ`` where ``θ`` represents the twist ensures that the result no longer depends on the order of evaluation.

Reviewer 2: ```suggestion
This modified map ``η′ := η ∘ θ`` where ``θ`` represents the twist ensures that the result no longer depends on the order of evaluation.
```

> In particular, one can show that any time two tensors would swap places, we would simultaneously exchange one evaluation map ``ϵ`` for a coevaluation ``̃η``, while also incurring a twist ``θ`` such that both cancel out.

Reviewer: ```suggestion
In particular, one can show that any time two tensor legs would swap places, we would simultaneously exchange one evaluation map ``ϵ`` for a coevaluation ``η′``, while also incurring a twist ``θ`` such that both cancel out.
```

To make this concrete, we show how our previous example now leads to a unique result:

```@example fermioncontraction
function fermion_mul(A, B)
    return A * twist(B, findall(isdual, codomain(B).spaces))
end

# order I:
AX = fermion_mul(repartition(A, 2, 1),  X)
AXB = fermion_mul(repartition(AX, 1, 2) , B)

# order II:
AB = fermion_mul(permute(A, ((1, 2), (3,))), permute(B, ((2,), (1, 3))))
ABX = repartition(fermion_mul(permute(AB, ((1, 4), (2, 3))), repartition(X, 2, 0)), 1, 1)

AXB ≈ ABX
```

This is the so-called **supertrace** formalism, and is effectively what `@tensor` ends up implementing for fermionic contractions.
For more details about this formalism, we refer to [^Mortier].

```@example fermioncontraction
# @tensor
@tensor result[-1; -2] := A[-1; 1 2] * X[1; 3] * B[3 2; -2]

AXB ≈ result
```

### (Non)-unitarity

> While this modified ``̃η`` solves the issues related to contractions, it does come at a cost.

Reviewer: ```suggestion
While this modified ``η′`` solves the issues related to contractions, it does come at a cost.
```
